// zoom-bot-sdk/src/plugins/sendMessage.js
import utils from '../utils/index';
import api from '../services/api';

let combineBody = function(body, header) {
  header = header || {
    text: 'text',
    style: {
      color: '#000000',
      bold: true,
      itatic: false
    }
  };
  if (!Array.isArray(body)) {
    body = [body];
  }
  let out = { head: header, body };
  return out;
};

let sendMessage = function(
  access_token,
  to_jid,
  robot_jid,
  account_id,
  body,
  header
) {
  return new Promise((resolve, reject) => {
    let url = api.sendMessage.get('url');
    let obj = {
      robot_jid,
      to_jid,
      account_id
    };
    // try{
    if (typeof body === 'string') {
      body = JSON.parse(body);
    }
    if (header && typeof header === 'string') {
      header = JSON.parse(header);
    }
    // }
    // catch(e){
    //   reject(e);
    //   return;
    // }

    let sendBody = Object.assign(obj, { content: combineBody(body, header) });

    let method = api.sendMessage.get('method');
    utils
      .request({
        url,
        method,
        body: sendBody,
        headers: {
          Authorization: `Bearer ${access_token}`
        }
      })
      .then(response => {
        let { body } = response;
        resolve(body);
      })
      .catch(error => {
        reject(error);
      });
  });
};

export default sendMessage;
