// zoom-bot-sdk/production/plugins/sendMessage.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _Utils = require("Utils");

var _Utils2 = _interopRequireDefault(_Utils);

var _api = require("Services/api");

var _api2 = _interopRequireDefault(_api);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

let combineBody = function (body, header) {
  header = header || {
    text: 'text',
    style: {
      color: '#000000',
      bold: true,
      itatic: false
    }
  };

  if (!Array.isArray(body)) {
    body = [body];
  }

  let out = {
    head: header,
    body
  };
  return out;
};

let sendMessage = function (access_token, to_jid, robot_jid, account_id, body, header) {
  return new Promise((resolve, reject) => {
    let url = _api2.default.sendMessage.get('url');

    let obj = {
      robot_jid,
      to_jid,
      account_id
    }; // try{

    if (typeof body === 'string') {
      body = JSON.parse(body);
    }

    if (header && typeof header === 'string') {
      header = JSON.parse(header);
    } // }
    // catch(e){
    //   reject(e);
    //   return;
    // }


    let sendBody = Object.assign(obj, {
      content: combineBody(body, header)
    });

    let method = _api2.default.sendMessage.get('method');

    _Utils2.default.request({
      url,
      method,
      body: sendBody,
      headers: {
        Authorization: `Bearer ${access_token}`
      }
    }).then(response => {
      let {
        body
      } = response;
      resolve(body);
    }).catch(error => {
      reject(error);
    });
  });
};

exports.default = sendMessage;
module.exports = exports.default;
